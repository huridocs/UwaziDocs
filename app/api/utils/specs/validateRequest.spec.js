import Joi from 'joi';

import middleware from '../validateRequest.js';

describe('validateRequest', () => {
  let next;
  let res;
  const req = {};

  beforeEach(() => {
    next = jest.fn();
    res = {
      json: jest.fn(),
      status: jest.fn(),
      error: jest.fn(),
    };
  });

  describe('using JOI schema', () => {
    it('should call next when body or query is valid', () => {
      const schema = Joi.object().keys({
        prop1: Joi.string(),
        prop2: Joi.number(),
      });

      req.body = { prop1: 'string', prop2: 25 };
      middleware(schema)(req, res, next);
      expect(next).toHaveBeenCalledWith();

      req.query = { prop1: 'string', prop2: 25 };
      middleware(schema, 'query')(req, res, next);
      expect(next).toHaveBeenCalledWith();
    });

    it('should call next with the error', () => {
      const schema = Joi.object().keys({
        prop1: Joi.string(),
        prop2: Joi.number(),
      });

      req.body = { prop1: 55, prop2: 'must be number' };
      middleware(schema)(req, res, next);

      expect(next.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('using AJV schema', () => {
    it('should call next when body or query is valid', async () => {
      const schema = {
        type: 'object',
        properties: {
          body: {
            type: 'object',
            properties: {
              prop1: { type: 'string' },
              prop2: { type: 'number' },
            },
          },
          query: {
            type: 'object',
            properties: {
              prop1: { type: 'string' },
              prop2: { type: 'number' },
            },
          },
        },
      };

      req.body = { prop1: 'string', prop2: 25 };
      await middleware(schema)(req, res, next);
      expect(next).toHaveBeenCalledWith();

      next.mockReset();
      req.query = { prop1: 'string', prop2: 25 };
      await middleware(schema)(req, res, next);
      expect(next).toHaveBeenCalledWith();
    });

    it('should call next with the error', async () => {
      const schema = {
        type: 'object',
        properties: {
          body: {
            type: 'object',
            properties: {
              prop1: { type: 'string' },
              prop2: { type: 'number' },
            },
          },
        },
      };

      req.body = { prop1: 55, prop2: 'must be number' };
      await middleware(schema)(req, res, next);
      const expected = expect.objectContaining({
        message: 'validation failed',
        errors: [
          expect.objectContaining({ message: 'must be string' }),
          expect.objectContaining({ message: 'must be number' }),
        ],
        code: 400,
      });
      expect(next.mock.calls[0]).toEqual([expected]);
    });
  });
});
