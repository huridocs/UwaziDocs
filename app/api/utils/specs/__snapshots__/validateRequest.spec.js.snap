// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRequest using AJV schema should call next with the error 1`] = `
Array [
  Object {
    "ajv": true,
    "code": 400,
    "errors": Array [
      Object {
        "dataPath": ".body.prop1",
        "keyword": "type",
        "message": "should be string",
        "params": Object {
          "type": "string",
        },
        "schemaPath": "#/properties/body/properties/prop1/type",
      },
      Object {
        "dataPath": ".body.prop2",
        "keyword": "type",
        "message": "should be number",
        "params": Object {
          "type": "number",
        },
        "schemaPath": "#/properties/body/properties/prop2/type",
      },
    ],
    "message": "validation failed",
    "validation": true,
  },
]
`;

exports[`validateRequest using JOI schema should call next with the error 1`] = `
Array [
  Object {
    "code": 400,
    "message": "ValidationError: child \\"prop1\\" fails because [\\"prop1\\" must be a string]",
  },
]
`;
